File: tmlang-go-compiler/parser.go
package main

import (
	"errors"
	"fmt"
)

type IntermediateRepresention struct {
	Meta struct {
		Start  string
		Accept string
		Reject string
	}

	Macros map[string][]Transition

	Main []Transition
}

type Target struct {
	Type   string // CALL or GOTO or RETURN
	Name   string // State name
	Return string // CALL <macro> -> q0         q0 is Return state
}

type Transition struct {
	Src    string
	Read   string
	Write  string
	Dir    string
	Target Target
}

type Parser struct {
	Tokens       []Token
	Position     int
	CurrentToken Token
	IR           IntermediateRepresention
}

func (parser *Parser) initParser(_tokens []Token) {
	parser.Tokens = _tokens
	parser.Position = 0
	if len(_tokens) > 0 {
		parser.CurrentToken = parser.Tokens[0]
	} else {
		parser.CurrentToken = Token{TypeOfToken: EOF}
	}
	parser.IR = IntermediateRepresention{}
	parser.IR.Macros = make(map[string][]Transition)
}

func (parser *Parser) advance() {
	parser.Position += 1
	if parser.Position < len(parser.Tokens) {
		parser.CurrentToken = parser.Tokens[parser.Position]
	} else {
		parser.CurrentToken = Token{}
	}
}

func (parser *Parser) consume(tokenType TokenType) (string, error) {
	if !parser.CurrentToken.isNil() && parser.CurrentToken.TypeOfToken == tokenType {
		currentTextValue := parser.CurrentToken.Value
		parser.advance()
		return currentTextValue, nil
	} else {
		currentTextValue := parser.CurrentToken.Value
		if parser.CurrentToken.isNil() {
			currentTextValue = "EOF"
		}

		return "", fmt.Errorf("Syntax Error: Expected %s but got %s at Line %d", tokenType, currentTextValue, parser.CurrentToken.Line)

	}
}

func (parser *Parser) parseConfig() error {
	_, err := parser.consume(SECTION) // Handle error
	if err != nil {
		return err
	}

	for i := 0; i < 3; i++ {

		configKeyword, err := parser.consume(KEYWORD)

		if err != nil {
			return err
		}

		configIdentifier, err := parser.consume(ID)

		if err != nil {
			return err
		}

		switch configKeyword {
		case "START:":
			parser.IR.Meta.Start = configIdentifier
		case "ACCEPT:":
			parser.IR.Meta.Accept = configIdentifier
		case "REJECT:":
			parser.IR.Meta.Reject = configIdentifier
		}
	}

	return nil
}

func (parser *Parser) parseMacros() error {
	if _, err := parser.consume(SECTION); err != nil { // Move parser to next token after "MACROS:", DEF <name>:
		return err
	}

	for parser.CurrentToken.TypeOfToken != SECTION {
		if _, err := parser.consume(KEYWORD); err != nil {
			return err
		}
		macroIdentifier, err := parser.consume(ID)
		if err != nil {
			return err
		}
		if _, err := parser.consume(COLON); err != nil {
			return err
		}
		var transitions []Transition
		for parser.CurrentToken.TypeOfToken == ID {
			transition, err := parser.parseTransition()

			if err != nil {
				return err
			}

			transitions = append(transitions, transition)
		}

		parser.IR.Macros[macroIdentifier] = transitions

	}

	return nil
}

func (parser *Parser) parseMain() error {
	_, err := parser.consume(SECTION) // Handle error
	if err != nil {
		return err
	}

	for parser.CurrentToken.TypeOfToken != EOF {
		transtion, err := parser.parseTransition()
		if err != nil {
			return err
		}
		parser.IR.Main = append(parser.IR.Main, transtion)
	}
	return nil
}

func (parser *Parser) parseTransition() (Transition, error) { // Parses main and macros transitions | q0, 1 -> 1, R, q0

	srcIdentifier, err := parser.consume(ID)
	if err != nil {
		return Transition{}, err

	}

	if _, err := parser.consume(COMMA); err != nil {
		return Transition{}, err

	}

	readSymbol, err := parser.consume(SYMBOL)
	if err != nil {
		return Transition{}, err

	}

	if _, err := parser.consume(ARROW); err != nil {
		return Transition{}, err

	}

	writeSymbol, err := parser.consume(SYMBOL)
	if err != nil {
		return Transition{}, err

	}

	if _, err := parser.consume(COMMA); err != nil {
		return Transition{}, err

	}

	direction, err := parser.consume(DIRECTION)
	if err != nil {
		return Transition{}, err

	}
	if _, err := parser.consume(COMMA); err != nil {
		return Transition{}, err

	}

	var target Target

	if parser.CurrentToken.TypeOfToken == KEYWORD { // CALL Keyword

		kw, err := parser.consume(KEYWORD)
		if err != nil {
			return Transition{}, err
		}
		switch kw {
		case "CALL":
			macroIdentifier, err := parser.consume(ID) // CALL move_to_end -> q1

			if err != nil {
				return Transition{}, err

			}

			if _, err := parser.consume(ARROW); err != nil {
				return Transition{}, err

			}
			returnStateIdentifier, err := parser.consume(ID)
			if err != nil {
				return Transition{}, err
			}

			target.Type = "CALL"
			target.Name = macroIdentifier
			target.Return = returnStateIdentifier

		case "RETURN":
			target.Type = "RETURN"
		}
	} else { // regular, q0, 0 -> 0, q1
		returnStateIdentifier, err := parser.consume(ID)

		if err != nil {
			return Transition{}, err

		}

		target.Type = "GOTO"
		target.Name = returnStateIdentifier
	}

	return Transition{
		srcIdentifier,
		readSymbol,
		writeSymbol,
		direction,
		target,
	}, nil

}

func (parser *Parser) parse() (IntermediateRepresention, error) {

	if parser.CurrentToken.TypeOfToken == SECTION && parser.CurrentToken.Value == "CONFIG:" {
		if err := parser.parseConfig(); err != nil {
			return IntermediateRepresention{}, err
		}
	} else {
		return parser.IR, errors.New("Program must contain a CONFIG section")
	}

	if parser.CurrentToken.TypeOfToken == SECTION && parser.CurrentToken.Value == "MACROS:" {
		parser.parseMacros()
	} // Macros are optinal

	if parser.CurrentToken.TypeOfToken == SECTION && parser.CurrentToken.Value == "MAIN:" {
		if err := parser.parseMain(); err != nil {
			return IntermediateRepresention{}, err
		}
	} else {
		return parser.IR, errors.New("Program must contain a MAIN section")
	}

	return parser.IR, nil
}

File: compiler/parser.py
from lexer import TokenType

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0
        self.current_token = self.tokens[0]
        
        # The Output Structure (Intermediate Representation)
        self.ir = {
            "meta": {},      # Stores START, ACCEPT, REJECT
            "macros": {},    # Stores macro definitions
            "main": []       # Stores main logic
        }

   
    def advance(self):
        self.pos += 1
        if self.pos < len(self.tokens):
            self.current_token = self.tokens[self.pos]
        else:
            self.current_token = None

   
    def consume(self, token_type):
        if self.current_token and self.current_token.type == token_type:
            val = self.current_token.value
            self.advance()
            return val
        else:
            current_val = self.current_token.type if self.current_token else "EOF"
            raise Exception(f"Syntax Error: Expected {token_type} but got {current_val} at Line {self.current_token.line if self.current_token else 'End'}")

    # Main Parser
    def parse(self):
       
        if self.current_token.type == TokenType.SECTION and self.current_token.value == "CONFIG:":
            self.parse_config()
        else:
            raise Exception("Syntax Error: Code must start with 'CONFIG:' section")
        
       
        if self.current_token.type == TokenType.SECTION and self.current_token.value == "MACROS:":
            self.parse_macros()

        if self.current_token.type == TokenType.SECTION and self.current_token.value == "MAIN:":
            self.parse_main()
        else:
            raise Exception("Syntax Error: Missing 'MAIN:' section")
            
        return self.ir

    
    def parse_config(self):
        print("Parsing Config...")
        self.consume(TokenType.SECTION) # Expect 'CONFIG:'
        
        # Expects 3 specific lines: START, ACCEPT, REJECT
        for _ in range(3):
            key = self.consume(TokenType.KEYWORD) # e.g. 'START:'
            val = self.consume(TokenType.ID)      # e.g. 'q0'
            
            if key == "START:": self.ir["meta"]["start"] = val
            elif key == "ACCEPT:": self.ir["meta"]["accept"] = val
            elif key == "REJECT:": self.ir["meta"]["reject"] = val

    def parse_macros(self):
        print("Parsing Macros...")
        self.consume(TokenType.SECTION)
 
        # Loop until we hit the 'MAIN:' section

        while self.current_token.type != TokenType.SECTION:
            self.consume(TokenType.KEYWORD) # Expect 'DEF'
            macro_name = self.consume(TokenType.ID)
            self.consume(TokenType.COLON)
            
            transitions = []
            # Keep parsing transitions until we hit another DEF or the MAIN section
            # We check if current token is an ID (start of transition) 
            while self.current_token.type == TokenType.ID:
                transitions.append(self.parse_transition())
            
            self.ir["macros"][macro_name] = transitions

    def parse_main(self):
        print("Parsing Main Logic...")
        self.consume(TokenType.SECTION)

        while self.current_token.type != TokenType.EOF:
            self.ir["main"].append(self.parse_transition())

    # --- The Core Logic: Parsing a single line ---
    def parse_transition(self):
        # Source State
        src = self.consume(TokenType.ID)
        self.consume(TokenType.COMMA)
        
        # Read Symbol
        read = self.consume(TokenType.SYMBOL)
        self.consume(TokenType.ARROW)
        
        # Write Symbol
        write = self.consume(TokenType.SYMBOL)
        self.consume(TokenType.COMMA)
        
        # Direction
        direction = self.consume(TokenType.DIRECTION)
        self.consume(TokenType.COMMA)
        
        # Target (The Logic Change is Here)
        target = {}
        
        if self.current_token.type == TokenType.KEYWORD:
            kw = self.consume(TokenType.KEYWORD)
            
            if kw == "CALL":
                # Logic: CALL <macro_name> -> <return_state>
                macro_name = self.consume(TokenType.ID)
                
                # Check for the return arrow
                self.consume(TokenType.ARROW)
                ret_state = self.consume(TokenType.ID)
                
                target = {
                    "type": "CALL", 
                    "name": macro_name, 
                    "ret": ret_state 
                }
                
            elif kw == "RETURN":
                # Logic: RETURN (Inside a macro, just ends)
                target = {"type": "RETURN"}
                
        else:
            # Logic: <state_name> (Standard GOTO)
            target_name = self.consume(TokenType.ID)
            target = {"type": "GOTO", "name": target_name}

        # Return the Transition Object
        return {
            "src": src,
            "read": read,
            "write": write,
            "dir": direction,
            "target": target
        }